### PR TITLE
Docstrings

### PR DESCRIPTION
**Pull request by opotowsky**
Probably needs editing

### CODE REVIEW

File: genapi.py [hunk 1]
```diff
@@ -877,6 +880,7 @@ def typesystem_pxd(ts, ns):
 DBTYPES_JS_URL = 'http://fuelcycle.org/arche/dbtypes.js'
 
 def parse_args(argv):
+    """Parses typesystem arguments for code generation."""
```
Comment (opotowsky) [inline genapi.py]: Do the arguments here have to be written about in a parameters section in the docstring?

File: cymetric/cyclus.pyx [hunk 2]
```diff
@@ -46,6 +47,24 @@ cdef class _Datum:
     #        free(self.ptx)
 
     def add_val(self, const char* field, value, shape=None, dbtype=cpp_typesystem.BLOB):
+        """Adds Datum values for type conversion.
```
Comment (scopatz) [inline cymetric/cyclus.pyx]: The 'for type conversion' here doesn't make a whole lot of sense.  Though this does do type conversion, what is important is that it adds a value to the current record.

File: cymetric/cyclus.pyx [hunk 3]
```diff
@@ -46,6 +47,24 @@ cdef class _Datum:
...
+        value : any data type
+            Value in table column.
+        shape : int
+            Length of value.
+        dbtype : cpp data type 
+            Data type as defined by cyclus typesystem
+
+        Returns
+        -------
+        self : any data type
```
Comment (scopatz) [inline cymetric/cyclus.pyx]: self is always a Datumn, not any data type.

File: cymetric/cyclus.pyx [hunk 4]
```diff
@@ -46,6 +47,24 @@ cdef class _Datum:
...
+        """Adds Datum values for type conversion.
+
+        Parameters
+        ----------
+        field : pointer to char/str
+            The column name.
+        value : any data type
+            Value in table column.
+        shape : int
+            Length of value.
```
Comment (scopatz) [inline cymetric/cyclus.pyx]: Shape is actually a list or tuple of ints.

File: cymetric/cyclus.pyx [hunk 5]
```diff
@@ -46,6 +47,24 @@ cdef class _Datum:
     #        free(self.ptx)
 
     def add_val(self, const char* field, value, shape=None, dbtype=cpp_typesystem.BLOB):
+        """Adds Datum values for type conversion.
+
+        Parameters
+        ----------
+        field : pointer to char/str
+            The column name.
+        value : any data type
```
Comment (scopatz) [inline cymetric/cyclus.pyx]: `object` is the correct Python terminology for any data type, though this is not incorrect.

File: cymetric/execution.py [hunk 6]
```diff
@@ -30,33 +30,41 @@ def __init__(self, name):
         self.conds = []
 
     def __call__(self, *args, **kwargs):
+        """Checks that column name is callable."""
```
Comment (scopatz) [inline cymetric/execution.py]: I would say something like "stub function that prevents ColumnProxies from being called."

File: cymetric/execution.py [hunk 7]
```diff
@@ -30,33 +30,41 @@ def __init__(self, name):
...
         raise TypeError('ColumnProxy object {0!r} is not callable'.format(self.name))
 
     def __getitem__(self, *args, **kwargs):
+        """Checks that column name can be indexed."""
```
Comment (scopatz) [inline cymetric/execution.py]: Same here.

File: cymetric/execution.py [hunk 8]
```diff
@@ -165,28 +177,36 @@ def __getitem__(self, key):
...
     def values(self):
+        """Returns copy of metric values."""
         return self._ctx.values()
 
     def items(self):
+        """Returns copy of metrics in (key, value) form."""
         return self._ctx.items()
 
     def __del__(self):
+        """Writes out data to db"""
```
Comment (scopatz) [inline cymetric/execution.py]: I would say, "closes db, flushing remaining buffers. "

File: cymetric/metrics.py [hunk 9]
```diff
@@ -41,6 +41,17 @@ def name(self):
 
 
 def _genmetricclass(f, name, depends, scheme):
+    """Metric class requirements.
```
Comment (scopatz) [inline cymetric/metrics.py]: I would say "create a new metric class with a given name, dependencies, and schema."

File: cymetric/metrics.py [hunk 10]
```diff
@@ -52,9 +63,11 @@ class Cls(Metric):
         __doc__ = inspect.getdoc(f)
 
         def __init__(self, db):
+            """Constructor for metric object in database."""
             super(Cls, self).__init__(db)
 
         def __call__(self, series, conds=None, known_tables=None, *args, **kwargs):
+            """Customize metric instance with parameters."""
```
Comment (scopatz) [inline cymetric/metrics.py]: Not really.  This computes the metric for a given input data and conditions.

File: cymetric/metrics.py [hunk 11]
```diff
@@ -68,6 +81,7 @@ def __call__(self, series, conds=None, known_tables=None, *args, **kwargs):
 
 
 def metric(name=None, depends=NotImplemented, schema=NotImplemented):
+    """Checks for existence of metric."""
```
Comment (scopatz) [inline cymetric/metrics.py]: This is a decorator that creates a metric class from a function (or class)

File: cymetric/root_metrics.py [hunk 12]
```diff
@@ -34,6 +39,7 @@ def __call__(self, conds=None, *args, **kwargs):
 
 
 def root_metric(obj=None, name=None, schema=None, *args, **kwargs):
+    """Checks for existence of root metric."""
```
Comment (scopatz) [inline cymetric/root_metrics.py]: Same decorator idea as with `metric()`. Please see "Effective Computation in Physics" for a description of decorators if you need one :).

File: genapi.py (hunk 1 above)
Comment (scopatz) [inline genapi.py]: Nope.

File: cymetric/root_metrics.py (hunk 12 above)
Comment (opotowsky) [inline cymetric/root_metrics.py]: I need to stop buy fancy cheese with all my textbook allowance! ;-)

Comment (scopatz) [inline cymetric/root_metrics.py]: I see that in America's dairyland cheese is more important than knowledge :)

**Comment by opotowsky** [comment]
@scopatz I'm all done :-)

File: cymetric/cyclus.pyx [hunk 13]
```diff
@@ -46,6 +47,23 @@ cdef class _Datum:
...
+        value : object
+            Value in table column.
+        shape : list or tuple of ints
+            Length of value.
+        dbtype : cpp data type 
+            Data type as defined by cyclus typesystem
+
+        Returns
+        -------
+        self : datum
```
Comment (scopatz) [inline cymetric/cyclus.pyx]: Capital D Datum, since you are referring to a class name.

**Comment by scopatz** [comment]
One last teensy tiny issue.

**Comment by opotowsky** [comment]
ddddDone!

**Pull request merged**